src/prefect/transactions.py

```diff
@@ -414,24 +438,221 @@ def rollback(self) -> bool:
+                )
+
+    async def rollback(self) -> bool:
+        if self.state in [TransactionState.ROLLED_BACK, TransactionState.COMMITTED]:
+            return False
+
+        try:
+            for hook in reversed(self.on_rollback_hooks):
+                await self.run_hook(hook, "rollback")
+
+            self.state: TransactionState = TransactionState.ROLLED_BACK
+
+            for child in reversed(self.children):
+                if isinstance(child, AsyncTransaction):
+                    await child.rollback()
+                else:
+                    child.rollback()
+
+            return True
+        except Exception:
+            if self.logger:
+                self.logger.exception(
+                    f"An error was encountered while rolling back transaction {self.key!r}",
+                    exc_info=True,
+                )
+            return False
+        finally:
+            if (
+                self.store
+                and self.key
+                and self.isolation_level == IsolationLevel.SERIALIZABLE
+            ):
+                self.logger.debug(f"Releasing lock for transaction {self.key!r}")
+                self.store.release_lock(self.key)
+
+    async def __aenter__(self) -> Self:
+        self.prepare_transaction()
+        await self.begin()
+        self._token = self.__var__.set(self)
+        return self
+
+    async def __aexit__(self, *exc_info: Any) -> None:
+        exc_type, exc_val, _ = exc_info
+        if not self._token:
+            raise RuntimeError(
+                "Asymmetric use of context. Context exit called without an enter."
+            )
+        if exc_type:
+            await self.rollback()
+            await self.reset()
+            raise exc_val
+
+        if self.commit_mode == CommitMode.EAGER:
+            await self.commit()
+
+        # if parent, let them take responsibility
+        if self.get_parent():
+            await self.reset()
+            return
+
+        if self.commit_mode == CommitMode.OFF:
+            # if no one took responsibility to commit, rolling back
+            # note that rollback returns if already committed
+            await self.rollback()
+        elif self.commit_mode == CommitMode.LAZY:
+            # no one left to take responsibility for committing
+            await self.commit()
+
+        await self.reset()
+
+    def __enter__(self) -> NoReturn:
+        raise NotImplementedError(
+            "AsyncTransaction does not support the `with` statement. Use the `async with` statement instead."
+        )
+
+    def __exit__(self, *exc_info: Any) -> NoReturn:
+        raise NotImplementedError(
+            "AsyncTransaction does not support the `with` statement. Use the `async with` statement instead."
+        )
 
 
-def get_transaction() -> Optional[Transaction]:
-    return Transaction.get_active()
+def get_transaction() -> BaseTransaction | None:
```

Comment:
just an idea: maybe `get_active_transaction` or `get_current_transaction`?

```suggestion
def get_transaction() -> BaseTransaction | None:
```

Reply:
Maybe, but I think deprecating and adding a new version should be in a different PR.